Merge MCP Server feature branch changes into integration

Azquelt opened the pull request:
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

All changes in this PR have been reviewed when they were merged into the `feature-mcp-server` branch.

IssacAX123 committed:
handles complex generics

IssacAX123 committed:
handles generic in tool param

fixes fat test for generics in params

restructures to resolve generics in toJSONSchemaObject

breaks down geenric map creation

custom generic exception

Refactor how we create type variable mappings

Consistently create a map of type variables to values for every schema
(even when the map will be empty).

Keep the genericMapStack state private to the SchemaGenerationContext
and instead provide methods to mutate and use it safely.

adds generics to defs

MCP schema test for reusing a generic class

Test reusing a generic class from two different locations but with the
same eventual values being provided for its type arguments.

MCP Generic Schema generation fixes

- Use the raw type in the TypeKey for a parameterized type
- Get the defs name correctly
- push and pop types into the context and resolve type variables when
  calculating class frequency

MCP Schema test for generic array

generic exception message code change

generic exception message code change

rebase wip

devalibm committed:
Refactor session deletion

devalibm committed:
Merge pull request #33326 from devalibm/delete-session

Delete session

devalibm committed:
Adds constants in a seperate and reusable class

devalibm committed:
Improves code readability and move constants to existing class

devalibm committed:
Removes McpConstants

devalibm committed:
Merge pull request #33360 from devalibm/const-issue

Puts all constants under one class and lets developers to reuse instead of declaring it everytime

Azquelt committed:
Update Javadoc on a few MCP classes

Azquelt committed:
Merge pull request #33367 from Azquelt/mcp-javadoc

Update Javadoc on a few MCP classes

Azquelt committed:
Change MCP tool definitions to use a handler

This is preparatory work for supporting tools registered
programatically without an annotated method.

- Add classes from quarkus-mcp: FeatureManager, ToolManager
- Split out the fields from ToolMetadata which are needed to call the
  method into MethodMetadata
- Create handlers for calling a bean method: SyncBeanMethodHandler,
  AsyncBeanMethodHandler. Much of the code here is moved from
  McpServlet.
- Move the Jsonb instance into McpCdiExtension since it's now needed to
  the method handler at startup
- Update McpServlet.callTool to call the tool handlers
- Remove requestId from CancellationImpl as it was unused
- Make McpToolCallParams return a map of arguments rather than an array
- Fix areas of the code which were using the Tool annotation directly
- Update test utilities which create ToolMetadata

habiblawal1 committed:
Handle MCP calls when CDI is inactive (#33379)

* fix: handle mcp calls when cdi is inactive

* chore: add mcp cdi inactive logs to nlsprops

* Update CDI inactive warning message

Co-authored-by: Andrew Rouse <anrouse@uk.ibm.com>

* Improve grammar for inactive CDI message in nlsprops

Co-authored-by: Ramkumar <88021034+ramkumar-k-9286@users.noreply.github.com>

* Improve grammar for inactive CDI message in nlsprops

Co-authored-by: Ramkumar <88021034+ramkumar-k-9286@users.noreply.github.com>

---------

Co-authored-by: Andrew Rouse <anrouse@uk.ibm.com>
Co-authored-by: Ramkumar <88021034+ramkumar-k-9286@users.noreply.github.com>

habiblawal1 committed:
Avoid tracing MCP Session ID (#33411)

* chore: store sessionId in McpSessionID class and hide sessionId in trace

* chore: extend redacted sessionid to 36 charaters

habiblawal1 committed:
fix: update inactiveCDI test message to match logs

habiblawal1 committed:
Merge pull request #33440 from habiblawal1/issue/33439-inactivecdi-test-bug

Fix inactivecdi test bug by updating message variable

Azquelt commented on dev/io.openliberty.mcp.internal/resources/io/openliberty/mcp/internal/resources/CWMCM.nlsprops:
```diff
@@ -126,6 +126,10 @@ CWMCM0016E.error.sending.response.exception=CWMCM0016E: Unexpected Server Error.
+# Used for server logging
+CWMCM0017W.cdi.inactive.warning=CWMCM0017W: CDI is not active. The MCP server cannot start. Verify that any MCP annotations are placed on methods of CDI beans that have an appropriate scope annotation (for example, @ApplicationScoped).
+CWMCM0017W.cdi.inactive.warning.explanation=CDI is not active which may occur if there are no CDI beans in the application.
+CWMCM0017W.cdi.inactive.warning.useraction=Check that any MCP annotations are placed on methods of CDI beans that have an appropriate scope annotation (for example, @ApplicationScoped or @RequestScoped). If your application archive contains a CDI extension, ensure that it also includes a beans.xml. If your application has a beans.xml, check that "bean-discovery-mode" is not "none".
```
ID Reviewed in #33379

Azquelt replied:
We removed this message anyway :upside_down_face:

Azquelt commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=a57c4ca4-0ae5-4392-92b4-aec1cd0b2c66) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

devalibm committed:
Updated internal implementation to match Quarkus-style typed RequestId handling

IssacAX123 committed:
Merge pull request #33333 from IssacAX123/33115-fix-generics-schema-bug

33115 fix generics schema bug

Azquelt committed:
Add MCP Async Tool error tests

Test all forms of error reporting from MCP Async tools.

IssacAX123 committed:
Adds ToolResponse Test (#33466)

* Adds Tool Resposne Test

* extracts json b to method parameter

* reverts back to qurkus tool response

* reverts javadoc

* reverts javadoc

* reverts tool response exception message

Azquelt committed:
Fix MCP Async tool exception handling

Handle exceptions correctly when the method returns a CompletionStage
which fails, rather than throwing the exception directly.

IssacAX123 committed:
updates test to use _meta (#33484)

devalibm committed:
Merge pull request #33461 from devalibm/mcp-33370

Updated internal implementation to match Quarkus-style typed RequestId handling

Azquelt committed:
Merge pull request #33384 from Azquelt/mcp-tool-handlers

Change MCP tool definitions to use a handler

martindrozdz committed:
MCP - Security Annotations implementation

Applies authorization checks for MCP resources (specifically calling and
listing tools).

This is only tested on stateful mode only for this PR.

We look for the annotation DenyAll, PermitAll, RolesAllowed for all Tool
methods and classes.

IssacAX123 committed:
33525 Fixes MetaKey to serialize to just String (#33531)

* converts Map<MetaKey, Object> to Map<String, Object> in responses

* adds tests for prefixes in _meta

* uses unreserved keys in _meta test

IssacAX123 committed:
33368 add meta parameter (#33566)

* injects Meta from request

* wip

* simple tests, moves _meta to outer params

* copyright header update

* package refactoring

* package trace update

habiblawal1 committed:
Improve description of MCP stateless mode (#33580)

* docs: improve description of MCP stateless mode

* addressed review comments

* describe mcp stateless mode as removing sticky sessions

habiblawal1 committed:
Change MCP CDI inactive warning to an info message (#33582)

* fix: change cdi inactive warning into an info message

* convert trace message into one not logged to user

martindrozdz committed:
Merge pull request #33516 from martindrozdz/feature-33118-SecAnnos

Security Annotations

Azquelt commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=8f75613d-c4b1-4f05-bfe5-e9563cc3ebb0) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

Azquelt commented on dev/io.openliberty.mcp.internal/resources/io/openliberty/mcp/internal/resources/CWMCM.nlsprops:
```diff
@@ -151,3 +151,11 @@ jsonrpc.extra.arguments=The following arguments were passed but were not found i
 # {0} = list of arguments
 jsonrpc.missing.arguments=The following arguments were expected by the method but were not provided: {0}.
+
+# Used for server logging
+# do not translate MCP
+# {0} = The qualified name of the tool
+# {1} = Parameter name
+CWMCM0018E.generic.arguments=CWMCM0018E: The {1} argument of the {0} MCP tool method contains unsupported components such as TypeVariable, Wildcard, GenericArrayType.
+CWMCM0018E.generic.arguments.explanation=MCP Tool arguments were provided but they are not supported. 
+CWMCM0018E.generic.arguments.useraction=Provide arguments that do not have generic components.
```
Message reviewed in #33333

martindrozdz replied:
Approving as requested in chat.

Azquelt commented on dev/io.openliberty.mcp.internal/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -2,4 +2,4 @@ mcpServer.name=Model Context Protocol (MCP) Server
 stateless.name=Stateless mode
-stateless.description=Stateless mode disables session tracking. This disabling enables easier clustering because the load balancer does not need to ensure that each session is served by a single backend server. However, it also disables MCP features that rely on session tracking or linking separate HTTP requests together, such as canceling a tool call
\ No newline at end of file
+stateless.description=Stateless mode disables MCP session tracking by the server. This mode removes the need for session affinity in a clustered environment. Otherwise, the same server handles requests from the same MCP session. Because no session data is stored on the servers, any server instance can handle any request, which simplifies load balancing and horizontal scaling. However, the statele ...
```
Reviewed in #33580

martindrozdz reviewed (commented):
Approving as requested in chat.

Azquelt committed:
Add MCP Tool tests for non-latin characters

Azquelt committed:
Use UTF-8 for MCP request and response

Azquelt committed:
Fix mcpServer-1.0 on EE 11

Azquelt committed:
Repeat MCP ToolTest on EE11 in LITE mode

Most tests only run on EE11 in FULL mode, but we should have one that
runs in LITE mode, which will catch most bundle packaging issues.

This is a bit of a hack until I find a better way to do it.

Azquelt commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=a69715ff-0838-4ecb-bd57-d61a0919650f) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._

Azquelt committed:
Merge pull request #33664 from Azquelt/mcp-charset

Use UTF-8 for MCP request and response

Azquelt committed:
Merge pull request #33667 from Azquelt/mcp-ee11-fix

Fix mcpServer-1.0 on EE 11

Azquelt commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=1c9bcae9-b10a-499e-b03b-5e607019d5b0) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._
